Enable support for Openstack cloud.

arcprabh opened the pull request:
Signed-off-by: arcprabh <arcprabh@in.ibm.com>

paigerube14 commented on run_kraken.py:
```diff
@@ -31,6 +32,8 @@ def get_node_scenario_object(node_scenario):
         return aws_node_scenarios()
     elif node_scenario['cloud_type'] == 'gcp':
         return gcp_node_scenarios()
+    elif node_scenario['cloud_type'] == 'openstackcloud':
```
Could we do ` 'openstack' in node_scenario['cloud_type']` instead, just to make the cloud type a little shorter

paigerube14 commented on docs/node_scenarios.md:
```diff
@@ -32,6 +32,10 @@ After creating the service account you'll need to enable the account using the f
 
+#### OPENSTACK
+
+**NOTE**: For clusters with OPENSTACK Cloud, make sure [OPENSTACK CLI](pip3 install python-openstackclient) is installed and then create and source the [OPENSTACK RC file](https://docs.openstack.org/newton/user-guide/common/cli-set-environment-variables-using-openstack-rc.html) from the server where Kraken runs.
```
I think we can add the openstack python client to the requirements

paigerube14 replied:
Also, can we move this section up above line 31 so all the clouds are together and maybe add another line in-between the end of openstack and the current line 31.

arcprabh commented:
New commit link is https://github.com/arcprabh/kraken/commit/92f81ed616b416c55ac602fedaa7ac98d9d4f1e2
Please review @paigerube14

paigerube14 commented on kraken/node_actions/openstack_node_scenarios.py:
```diff
@@ -0,0 +1,121 @@
+                openstack_node_ip = nodeaction.get_node_ip(node)
+                openstack_node_name = self.openstackcloud.get_openstack_nodename(openstack_node_ip)
+                self.openstackcloud.stop_instances(openstack_node_name)
+                self.openstackcloud.wait_until_running(openstack_node_name)
```
Should be `wait_until_stopped` instead of running

arcprabh replied:
Good catch.  I will change it.

paigerube14 commented:
In the _get_openstack_nodename_ function I ran into a problem 


`server_list = runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
        list_of_servers = server_list.split('\n')

        for item in list_of_servers:
            items = item.split("|")
            counter = 0
            for i in items:
                if i.strip() != "" and counter == 2:
                    node_name = i.strip()
                    counter += 1
                    continue
                item_list = i.split('=')
                if len(item_list) == 2 and item_list[-1].strip() == os_node_ip:
                    return node_name
                counter += 1`

paigerube14 commented:
In the get_openstack_nodename function I ran into a problem with getting the node name from the IP address. It seemed that in our OpenStack server list there were multiple instance matches when _greping_ for the IP address (see example response below) 

Ex.) runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
 server list response: | 48e050e7-225b-4942-805e-f744f2bf3f62 | _ug-bootstrap_                                         | ACTIVE  | ugd2932022589-rm94d-openshift=192.168.3.159, 10.0.102.170                     | rhcos-42.80.20191010.0                      | m1.xlarge      |
| a2ed6f85-5a84-4249-9eb8-c029541ec464 | _qe-pr-osp2-hzmvf-worker-0-87jks_                                       | ACTIVE  | qe-pr-osp2-hzmvf-openshift=192.168.3.15                                       | qe-pr-osp2-hzmvf-rhcos                      | m1.large       |

So the response from the function was the 2 node names concatenated and then when it tried to stop the node the OpenStack client couldn’t find a node with name “ug-bootstrapqe-pr-osp2-hzmvf-worker-0-87jks”

I coded up something to get past that point. Feel free to rewrite if there’s a better way 
`server_list = runcommand.invoke("openstack server list | grep %s" % (os_node_ip)) list_of_servers = server_list.split('\n')
    for item in list_of_servers:
        items = item.split("|")
        counter = 0
        for i in items:
            if i.strip() != "" and counter == 2:
                node_name = i.strip()
                counter += 1
                continue
            item_list = i.split('=')
            if len(item_list) == 2 and item_list[-1].strip() == os_node_ip:
                return node_name
            counter += 1`

I’m not sure if this depends on the set up of an Openshift cluster, but for me the node name and the name of the open stack instance are the same so the get ip address and get open stack node name are redundant. If this is not the case for your set up please keep the functions; just wanted to call it out that that’s how our test environments are set up. 

Are you also able to add in the terminate node code when you get a chance. 

Looking great though! Thanks for the contributions and quick code changes already

paigerube14 commented:
After changing those couple of things listed above I was able to succesfully start, stop and reboot nodes in Openstack :) 

Please be sure to run "tox" as well to correct all formatting issues.

arcprabh commented:
> In the get_openstack_nodename function I ran into a problem with getting the node name from the IP address. It seemed that in our OpenStack server list there were multiple instance matches when _greping_ for the IP address (see example response below)
> 
> Ex.) runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
> server list response: | 48e050e7-225b-4942-805e-f744f2bf3f62 | _ug-bootstrap_ | ACTIVE | ugd2932022589-rm94d-openshift=**192.168.3.15**9, 10.0.102.170 | rhcos-42.80.20191010.0 | m1.xlarge |
> | a2ed6f85-5a84-4249-9eb8-c029541ec464 | _qe-pr-osp2-hzmvf-worker-0-87jks_ | ACTIVE | qe-pr-osp2-hzmvf-openshift=**192.168.3.15** | qe-pr-osp2-hzmvf-rhcos | m1.large |
> 
> So the response from the function was the 2 node names concatenated and then when it tried to stop the node the OpenStack client couldn’t find a node with name “ug-bootstrapqe-pr-osp2-hzmvf-worker-0-87jks”
> 
> I coded up something to get past that point. Feel free to rewrite if there’s a better way
> 
> `
> server_list = runcommand.invoke("openstack server list | grep %s" % (os_node_ip))
> 
> ```
>  list_of_servers = server_list.split('\n')
> 
> for item in list_of_servers:
> 
>     items = item.split("|")
> 
>     counter = 0
> 
>     for i in items:
> 
>         if i.strip() != "" and counter == 2:
> 
>            node_name = i.strip()
> 
>             counter += 1
> 
>             continue
> 
>         item_list = i.split('=')
> 
>         if len(item_list) == 2 and item_list[-1].strip() == os_node_ip:
> 
>             return node_name
> 
>         counter += 1
> ```
> 
> `
> 
> I’m not sure if this depends on the set up of an Openshift cluster, but for me the node name and the name of the open stack instance are the same so the get ip address and get open stack node name are redundant. If this is not the case for your set up please keep the functions; just wanted to call it out that that’s how our test environments are set up.
> 
> Are you also able to add in the terminate node code when you get a chance.
> 
> Looking great though! Thanks for the contributions and quick code changes already

Tested this code and it works on my set up too.

paigerube14 commented:
LGTM

pruan-rht commented on kraken/node_actions/openstack_node_scenarios.py:
```diff
@@ -0,0 +1,133 @@
+from kraken.node_actions.abstract_node_scenarios import abstract_node_scenarios
+
+
+class OPENSTACKCLOUD:
```
@paigerube14 I think python [PEP8](https://www.python.org/dev/peps/pep-0008/#class-names) states that class names should be CamelCase.

pruan-rht commented on kraken/node_actions/openstack_node_scenarios.py:
```diff
@@ -0,0 +1,133 @@
+                   return node_name
+                counter += 1
+ 
+class openstack_node_scenarios(abstract_node_scenarios):
```
ditto

mffiedler commented:
@pruan-rht we can do a follow up PR if needed for clean up.  Merging to get the functionality in.

smalleni commented on kraken/node_actions/openstack_node_scenarios.py:
```diff
@@ -0,0 +1,133 @@
+
+    # Start the node instance
+    def start_instances(self,node):
+        runcommand.invoke("openstack server start %s" % (node))
```
Thought about actually using the openstack client instead of subprocessing commands?
